### PR TITLE
OVMM real world integration - first pass

### PR DESCRIPTION
**Pull request by cpaxton**
## Motivation and Context

Initial pass at OVMM integration based on existing object nav code from Devendra and Theo

## Changelog

- Bug fixes for getting the robot to work in the real world
- Switching from env to robot for interface - remove old environment configs
- Tuning parameters in stretch_objectnav and stretch_grasping
- Adding hierarchical agent and testing object nav + grasping handoff
- Tuning motion planning based on FMM
- Restricting frontier exploration so that we can find objects more reliably when obstacles are added all over the place
- Adding prototypes for sampling based planner - implementation still to-do
- Switching robot manipulation mode so that we can move base and grasp
- improving inverse kinematics code - adding multiple attempts with random initializations, plus expanding the template for ik solvers and fixing various test failures that resulted from bad initializations

## How Has This Been Tested

- testing in the real world
- test_ik function

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [x] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: projects/stretch_grasping/test_pick.py
Comment (exhaustin) [inline projects/stretch_grasping/test_pick.py:1]: Remove this file? It's just a debugging script, right?

File: projects/stretch_objectnav/rotate_in_place.py
Comment (exhaustin) [inline projects/stretch_objectnav/rotate_in_place.py:1]: Also remove these debugging scripts?

File: src/home_robot/home_robot/agent/objectnav_agent/objectnav_agent.py [hunk 1]
```diff
@@ -285,16 +291,24 @@ def _preprocess_obs(self, obs: Observations):
...
+        if (
+            "recep_goal" in obs.task_observations
+            and obs.task_observations["recep_goal"] is not None
+        ):
             recep_goal_category = torch.tensor(
                 obs.task_observations["recep_goal"]
             ).unsqueeze(0)
         goal_name = [obs.task_observations["goal_name"]]
+        print("object goal =", obs.task_observations["object_goal"])
+        print("recep goal =", obs.task_observations["recep_goal"])
```
Comment (zephirefaith) [inline src/home_robot/home_robot/agent/objectnav_agent/objectnav_agent.py:311]: Both of these will fault when these keys are not present. Since you are anyways checking for those keys above, move these under there?

File: src/home_robot/home_robot/agent/objectnav_agent/sampling_agent.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        self.voxel_map = SparseVoxelMap()
+
+    def reset(self):
+        """Clear information in the voxel map"""
+        self.reset_vectorized()
+        self.voxel_map.reset()
+        self.planner.reset()
+        self.episode_panorama_start_steps = self.panorama_start_steps
+
+    def act(self, obs: Observations) -> Tuple[DiscreteNavigationAction, Dict[str, Any]]:
```
Comment (zephirefaith) [inline src/home_robot/home_robot/agent/objectnav_agent/sampling_agent.py:38]: Just flagging that this is unfinished. Maybe raise an unimplementederror so you remember this is unfinished?

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 3]
```diff
@@ -86,15 +97,21 @@ def __init__(
         self.local_map_size = self.local_map_size_cm // self.resolution
         self.xy_resolution = self.z_resolution = map_resolution
```
Comment (zephirefaith) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:98]: There's a bunch of divisions here; maybe add defaults/check if anything is 0 and raise error?

File: src/home_robot/home_robot/navigation_planner/discrete_planner.py [hunk 4]
```diff
@@ -138,49 +145,110 @@ def plan(
...
-            if self.curr_obs_dilation_selem_radius > 1:
-                self.curr_obs_dilation_selem_radius -= 1
-                self.obs_dilation_selem = skimage.morphology.disk(
-                    self.curr_obs_dilation_selem_radius
-                )
+            print("Could not find a path to the high-level goal. Stopping.")
+            action = DiscreteNavigationAction.STOP
+            # TODO Calling the STOP action here will cause the agent to try grasping
+            #   we need different STOP_SUCCESS and STOP_FAILURE actions
+            breakpoint()
```
Comment (zephirefaith) [inline src/home_robot/home_robot/navigation_planner/discrete_planner.py:188]: remove the breakpoint? and the dead code underneath?

File: src/home_robot/home_robot/navigation_planner/fmm_planner.py [hunk 5]
```diff
@@ -167,3 +171,51 @@ def get_dist(sx, sy, scale, step_size):
                         ** 0.5,
                     )
         return mask
+
+    def _find_nearest_to_multi_goal(self, goal):
```
Comment (zephirefaith) [inline src/home_robot/home_robot/navigation_planner/fmm_planner.py:175]: type hints for both the methods

File: src/home_robot/home_robot/navigation_planner/rrt.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        self.state = state
+
+
+class RRTPlanner:
+    """Sampling-based planner."""
+
+    def __init__(self):
+        """Set up default params for interpolating states, sampling rates, etc."""
+
+    def solve(self, voxel_map: SparseVoxelMap, goal: BaseState):
```
Comment (zephirefaith) [inline src/home_robot/home_robot/navigation_planner/rrt.py:38]: raise not implemented error.

File: src/home_robot_hw/install.md
Comment (zephirefaith) [inline src/home_robot_hw/install.md:1]: Cool! Thanks for improving the installation documentation!

**Review by zephirefaith — CHANGES_REQUESTED** [review]

File: src/home_robot/home_robot/agent/objectnav_agent/sampling_agent.py (hunk 2 above)
Comment (cpaxton) [reply src/home_robot/home_robot/agent/objectnav_agent/sampling_agent.py:38]: done

File: src/home_robot/home_robot/navigation_planner/discrete_planner.py (hunk 4 above)
Comment (cpaxton) [reply src/home_robot/home_robot/navigation_planner/discrete_planner.py:188]: changed a bit but this is WIP

File: projects/stretch_grasping/utils/rotate_left.py
[the remark this answers is not in the document]
Comment (cpaxton) [reply projects/stretch_grasping/utils/rotate_left.py:1]: deleted

File: projects/stretch_grasping/test_pick.py
Comment (cpaxton) [reply projects/stretch_grasping/test_pick.py:1]: i want to keep this until IK is reliably working on this one robot

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py (hunk 3 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:98]: these are all presets and this is from devendra's code, would rather not change anything

File: src/home_robot/home_robot/navigation_planner/discrete_planner.py (hunk 4 above)
Comment (cpaxton) [reply src/home_robot/home_robot/navigation_planner/discrete_planner.py]: should be ok now, will just terminate

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 7]
```diff
@@ -67,13 +82,15 @@ def __init__(
              consider it as explored
             map_pred_threshold: number of depth points to be in bin to
              consider it as obstacle
+            must_explore_close: reduce the distance we need to get to things to make them work
```
Comment (theophilegervet) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:85]: @cpaxton Let's put this in the config?

Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:85]: I'm not so familiar with these configs, what do i need to do in order to do this?

File: src/home_robot_hw/home_robot_hw/env/stretch_pick_and_place_env.py [hunk 8]
```diff
@@ -114,59 +123,76 @@ def apply_action(self, action: Action, info: Optional[Dict[str, Any]] = None):
...
+        goal_obj = self.goal_options[goal_obj_idx]
+        self.current_goal_id = self.goal_options.index(goal_obj_idx)
+        self.current_goal_name = goal_obj
 
     def get_observation(self) -> Observations:
         """Get Detic and rgb/xyz/theta from this"""
-        rgb, depth, xyz = self.get_images(compute_xyz=True, rotate_images=True)
+        rgb, depth, xyz = self.robot.head.get_images(
+            compute_xyz=True,
+        )
```
Comment (zephirefaith) [inline src/home_robot_hw/home_robot_hw/env/stretch_pick_and_place_env.py:189]: Just double-checking that you've run this on the robot and PCDs looked fine with this argument change?

**Review by zephirefaith — APPROVED** [review]
Looks good. A few nits. Thank you for making the changes!

File: src/home_robot_hw/home_robot_hw/utils/grasping.py
Comment (exhaustin) [inline src/home_robot_hw/home_robot_hw/utils/grasping.py:1]: Probably okay to refactor later, but the functionality provided in this file is too complicated and not general enough to belong in utils.

File: src/home_robot/home_robot/motion/pinocchio_ik_solver.py [hunk 9]
```diff
@@ -80,21 +80,35 @@ def compute_fk(self, q) -> Tuple[np.ndarray, np.ndarray]:
...
+            num_attempts = 1
         desired_ee_pose = pinocchio.SE3(R.from_quat(quat).as_matrix(), pos)
         while True:
             pinocchio.forwardKinematics(self.model, self.data, q)
             pinocchio.updateFramePlacement(self.model, self.data, self.ee_frame_idx)
 
             dMi = desired_ee_pose.actInv(self.data.oMf[self.ee_frame_idx])
             err = pinocchio.log(dMi).vector
+            if verbose:
+                print(i, err)
```
Comment (exhaustin) [inline src/home_robot/home_robot/motion/pinocchio_ik_solver.py:111]: Add more info to the printout if we want this to be a feature and not just a debug print.
Maybe `print("[pinocchio_ik_solver] iter={i}; error={err}")`

File: src/home_robot/home_robot/navigation_planner/rrt.py
Comment (exhaustin) [inline src/home_robot/home_robot/navigation_planner/rrt.py:1]: Can we move this into a separate PR?

**Review by exhaustin — APPROVED** [review]
Added some minor comments, overall if the robot works and tests pass then looks good.
This PR is growing to be quite large and I'm leaning more towards a merge now fix later mentality, so approving to unblock.

File: src/home_robot_hw/home_robot_hw/utils/grasping.py
Comment (cpaxton) [reply src/home_robot_hw/home_robot_hw/utils/grasping.py:1]: yeah it should not be called "utils" - should be moved to something else since its our grasp planner

File: src/home_robot_hw/home_robot_hw/env/stretch_pick_and_place_env.py (hunk 8 above)
Comment (cpaxton) [reply src/home_robot_hw/home_robot_hw/env/stretch_pick_and_place_env.py:189]: yeah it looks fine

File: src/home_robot/home_robot/navigation_policy/object_navigation/objectnav_frontier_exploration_policy.py [hunk 10]
```diff
@@ -171,7 +175,11 @@ def explore_otherwise(self, map_features, goal_map, found_goal):
         frontier_map = (
             binary_dilation(frontier_map, self.select_border_kernel) - frontier_map
         )
+        return frontier_map
 
+    def explore_otherwise(self, map_features, goal_map, found_goal):
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply src/home_robot/home_robot/navigation_policy/object_navigation/objectnav_frontier_exploration_policy.py:180]: this is devendra's code, not sure why its marked as new here; i'd rather not touch it

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py (hunk 7 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:85]: Can we add this as an issue and tag me? I dont want to make these kinds of changes without hardware tests

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 11]
```diff
@@ -253,28 +276,67 @@ def _update_local_map_and_pose(
         batch_size, obs_channels, h, w = obs.size()
         device, dtype = obs.device, obs.dtype
         if camera_pose is not None:
-            tilt = pt.matrix_to_euler_angles(camera_pose[:, :3, :3], convention="YZX")[
-                :, -1
-            ]
+            # TODO: make consistent between sim and real
+            # hab_angles = pt.matrix_to_euler_angles(camera_pose[:, :3, :3], convention="YZX")
+            angles = pt.matrix_to_euler_angles(camera_pose[:, :3, :3], convention="ZYX")
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:281]: Hi Theo! Yes, I handled it [here](https://github.com/facebookresearch/home-robot/blob/b499c994bb3e5561819d0cc7583b6798b4ab9260/src/home_robot_sim/home_robot_sim/env/habitat_ovmm_env/habitat_ovmm_env.py#L89).

[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:281]: This TODO can be removed.

File: src/home_robot/home_robot/motion/pinocchio_ik_solver.py (hunk 9 above)
Comment (cpaxton) [reply src/home_robot/home_robot/motion/pinocchio_ik_solver.py]: done

File: src/home_robot/home_robot/motion/stretch.py [hunk 12]
```diff
@@ -613,16 +634,26 @@ def ik(self, pose, q0):
...
         """IK in manipulation mode. Takes in a 4x4 pose_query matrix in se(3) and initial
         configuration of the robot.
 
         By default move relative. easier that way.
         """
 
-        if q0 is None:
-            q0 = STRETCH_HOME_Q
+        if q0 is not None:
+            self._to_manip_format(q0)
```
[the remark this answers is not in the document]
Comment (cpaxton) [reply src/home_robot/home_robot/motion/stretch.py:649]: i'm nervous about changing this without real world tests, let's just let you fix it in your next slap PR
